src/VisualStudio/Core/Def/PdbSourceDocument/SourceLinkService.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.IO;
+using System.Reflection.PortableExecutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PdbSourceDocument;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.Debugger.Contracts.SourceLink;
+using Microsoft.VisualStudio.Debugger.Contracts.SymbolLocator;
+
+namespace Microsoft.VisualStudio.LanguageServices.PdbSourceDocument
+{
+    [Export(typeof(ISourceLinkService)), Shared]
+    internal class SourceLinkService : ISourceLinkService
+    {
+        private readonly IDebuggerSymbolLocatorService _debuggerSymbolLocatorService;
+        private readonly IDebuggerSourceLinkService _debuggerSourceLinkService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SourceLinkService(IDebuggerSymbolLocatorService debuggerSymbolLocatorService, IDebuggerSourceLinkService debuggerSourceLinkService)
+        {
+            _debuggerSymbolLocatorService = debuggerSymbolLocatorService;
+            _debuggerSourceLinkService = debuggerSourceLinkService;
+        }
+
+        public async Task<PdbFilePathResult?> GetPdbFilePathAsync(string dllPath, PEReader peReader, IPdbSourceDocumentLogger? logger, CancellationToken cancellationToken)
+        {
+            var hasCodeViewEntry = false;
+            uint timeStamp = 0;
+            CodeViewDebugDirectoryData codeViewEntry = default;
+            using var _ = ArrayBuilder<PdbChecksum>.GetInstance(out var checksums);
+            foreach (var entry in peReader.ReadDebugDirectory())
+            {
+                if (entry.Type == DebugDirectoryEntryType.PdbChecksum)
+                {
+                    var checksum = peReader.ReadPdbChecksumDebugDirectoryData(entry);
+                    checksums.Add(new PdbChecksum(checksum.AlgorithmName, checksum.Checksum));
+                }
+                else if (entry.Type == DebugDirectoryEntryType.CodeView && entry.IsPortableCodeView)
```

Comment:
Will this line need to change at all for windows pdbs?

Reply:
Potentially yes, though there could be multiple codeview entries and so we might leave this so that we prefer portable, but add another case. Not sure yet